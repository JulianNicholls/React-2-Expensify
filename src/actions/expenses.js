import database from '../firebase';

import { ADD_EXPENSE, REMOVE_EXPENSE, EDIT_EXPENSE, SET_EXPENSES } from './types';

// ADD EXPENSE

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database
      .ref('expenses')
      .push(expense)
      .then(ref => dispatch(addExpense({ id: ref.key, ...expense })));
  };
};

export const addExpense = expense => ({ type: ADD_EXPENSE, expense });

// REMOVE EXPENSE

export const startRemoveExpense = id => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => dispatch(removeExpense(id)));
  };
};

export const removeExpense = id => ({ type: REMOVE_EXPENSE, id });

// EDIT EXPENSE

export const startEditExpense = (id, updates) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .update(updates)
      .then(() => dispatch(editExpense(id, updates)));
  };
};

export const editExpense = (id, updates) => ({ type: EDIT_EXPENSE, id, updates });

// SET EXPENSES

export const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref('expenses')
      .once('value')
      .then(snapshot => {
        const expensesData = [];

        snapshot.forEach(item => {
          expensesData.push({ id: item.key, ...item.val() });
        });

        dispatch(setExpenses(expensesData));
      });
  };
};

export const setExpenses = expenses => ({ type: SET_EXPENSES, expenses });
