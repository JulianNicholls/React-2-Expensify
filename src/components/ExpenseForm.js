import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

class ExpenseForm extends React.Component {
  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    pickerFocused: false,
    error: ''
  };

  changeDescription = e => {
    const description = e.target.value;

    this.setState(() => ({ description }));
  };

  changeNote = e => {
    const note = e.target.value;

    this.setState(() => ({ note }));
  };

  changeAmount = e => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
      this.setState(() => ({ amount }));
  };

  onDateChange = createdAt => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ pickerFocused: focused }));
  };

  addExpense = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'You must enter a description and an amount' }));
    } else {
      this.setState(() => ({ error: '' }));

      this.props.handleSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p className="form-error">{this.state.error}</p>}
        <form onSubmit={this.addExpense} className="expense-form">
          <input
            value={this.state.description}
            type="text"
            placeholder="Description"
            autoFocus={true}
            onChange={this.changeDescription}
          />
          <input
            value={this.state.amount}
            type="text"
            placeholder="Amount"
            onChange={this.changeAmount}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.pickerFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            displayFormat={'ll'}
          />
          <textarea
            value={this.state.note}
            placeholder="Add an optional note for this expense"
            onChange={this.changeNote}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
