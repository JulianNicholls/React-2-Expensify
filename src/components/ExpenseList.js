import React from 'react';
import { connect } from 'react-redux';

import filteredExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = ({ expenses }) => (
  <div>{expenses.map(item => <ExpenseListItem key={item.id} {...item} />)}</div>
);

const mapStateToProps = state => ({
  expenses: filteredExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
