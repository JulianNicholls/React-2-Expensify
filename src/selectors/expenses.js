export default (expenses, { text, sortBy, startDate, endDate }) => {
  text = text.toLowerCase();

  const filtered = expenses.filter(({ description, createdAt }) => {
    const startDateMatch = startDate ? startDate <= createdAt : true;
    const endDateMatch = endDate ? endDate >= createdAt : true;
    const textMatch = text === '' || description.toLowerCase().includes(text);

    return startDateMatch && endDateMatch && textMatch;
  });

  // Sort the expenses, both date and amount are sorted into descending order.

  return filtered.sort((a, b) => {
    if (sortBy === 'date') return b.createdAt - a.createdAt;

    return b.amount - a.amount;
  });
};
