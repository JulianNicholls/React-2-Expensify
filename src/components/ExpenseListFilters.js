import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import * as actions from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    pickerFocused: null
  };

  onTextChange = e => this.props.setTextFilter(e.target.value);

  onSortChange = e =>
    e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = pickerFocused => this.setState(() => ({ pickerFocused }));

  render() {
    const {
      filters: { text, sortBy, startDate, endDate }
    } = this.props;

    return (
      <div>
        <input type="text" value={text} onChange={this.onTextChange} />
        <select value={sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={startDate}
          startDateId="sdid" // New requirement
          endDate={endDate}
          endDateId="edid" // New requirement
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.pickerFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          monthFormat={'MMM YY'}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

export default connect(
  mapStateToProps,
  actions
)(ExpenseListFilters);
