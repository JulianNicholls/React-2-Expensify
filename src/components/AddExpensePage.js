import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export const AddExpensePage = ({ addExpense, history }) => (
  <div className="container">
    <h1>Add Expense</h1>
    <ExpenseForm
      handleSubmit={expense => {
        addExpense(expense);

        history.push('/');
      }}
    />
  </div>
);

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ addExpense }, dispatch);
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
