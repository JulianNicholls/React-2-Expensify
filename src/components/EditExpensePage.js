import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import * as actions from '../actions/expenses';

export class EditExpensePage extends React.Component {
  handleSubmit = updated => {
    this.props.editExpense(this.props.expense.id, updated);

    this.props.history.push('/');
  };

  handleRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);

    this.props.history.push('/');
  };

  render() {
    return (
      <div className="container">
        <h2>Edit Expense</h2>
        <ExpenseForm expense={this.props.expense} handleSubmit={this.handleSubmit} />
        <button onClick={this.handleRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(item => item.id === props.match.params.id)
  };
};

export default connect(
  mapStateToProps,
  actions
)(EditExpensePage);
