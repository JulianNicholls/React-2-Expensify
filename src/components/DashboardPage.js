import React from 'react';

import ExpenseList from './ExpenseList';

const DashboardPage = () => (
  <div className="container">
    <h1>Dashboard</h1>
    <ExpenseList />
  </div>
);

export default DashboardPage;
