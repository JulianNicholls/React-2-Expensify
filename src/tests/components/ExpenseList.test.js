import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseList } from '../../components/ExpenseList';
import { testExpenses } from '../fixtures/expenses';

describe('ExpenseList component', () => {
  it('should render correctly WITH expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={testExpenses} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with NO expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);

    expect(wrapper).toMatchSnapshot();
  });
});
