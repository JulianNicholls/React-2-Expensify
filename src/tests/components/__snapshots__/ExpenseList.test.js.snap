// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList component should render correctly WITH expenses 1`] = `
<div>
  <ExpenseListItem
    amount={10000}
    createdAt={2000000000}
    description="rent"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={15000}
    createdAt={1800000000}
    description="rates"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={20000}
    createdAt={1600000000}
    description="water"
    id="3"
    key="3"
    note=""
  />
  <ExpenseListItem
    amount={25000}
    createdAt={1400000000}
    description="gas"
    id="4"
    key="4"
    note=""
  />
</div>
`;

exports[`ExpenseList component should render correctly with NO expenses 1`] = `<div />`;
